Framework/API/src/FileProperty.cpp

```diff
@@ -396,5 +401,55 @@ std::string FileProperty::convertExtension(const std::string &filepath) const {
   fullpath.setExtension(ext);
   return fullpath.toString();
 }
+
+/** Expand user variables in file path.
+ *  On UNIX, ~ is replaced by the user's home directory, if found.
+ *  On Windows, a combination of HOMEPATH and HOMEDRIVE will be used.
+ *  If the path contains no user variables, or expansion fails, the path is
+ *  returned unchanged, for errors to be dealt with by the calling function
+ *  @param filepath The path to expand
+ *  @return The expanded path
+ */
+std::string FileProperty::expandUser(const std::string &filepath) const {
+  auto start = filepath.begin();
+
+  if (*start != '~') // No user variables in filepath
+    return filepath;
+
+  // Position of the first slash after the variable
+  auto nextSlash = find_if(start, filepath.end(), &isSlash);
+
+  // UNIX-style home variable (ie ~/blah)
+  if (std::distance(start, nextSlash) == 1) {
+    char *home;
+
+    if (!(home = std::getenv("HOME")))
+      if (!(home = std::getenv("USERPROFILE")))
+        if (!(home = std::getenv("HOMEPATH")))
```

Comment:
It's always possible that a user could set those to empty values. In this case I would just expect us to fail gracefully with a reasonable error message. I would use the values as they are given and if the path is not correct then it should give the normal message that it couldn't find the file.

Does https://github.com/python/cpython/blob/2.7/Lib/ntpath.py#L285 help? This is how python does it. Similarly the `posixpath` module defines the rules for Unix.

Can I suggest we `ifdef` this and use different rules for Windows and Unix?